src/DecryptionFailureTracker.ts

```diff
@@ -176,21 +220,90 @@ export class DecryptionFailureTracker {
         }
 
         this.failures.set(eventId, failure);
-        if (this.visibleEvents.has(eventId) && !this.visibleFailures.has(eventId)) {
-            this.visibleFailures.set(eventId, failure);
-        }
     }
 
-    public removeDecryptionFailuresForEvent(e: MatrixEvent): void {
+    public removeDecryptionFailuresForEvent(e: MatrixEvent, nowTs: number): void {
         const eventId = e.getId()!;
-        this.failures.delete(eventId);
-        this.visibleFailures.delete(eventId);
+        const failure = this.failures.get(eventId);
+        if (failure) {
+            this.failures.delete(eventId);
+
+            const timeToDecryptMillis = nowTs - failure.ts;
+            if (timeToDecryptMillis < DecryptionFailureTracker.GRACE_PERIOD_MS) {
+                // the event decrypted on time, so we don't need to report it
+                return;
+            } else if (timeToDecryptMillis <= DecryptionFailureTracker.MAXIMUM_LATE_DECRYPTION_PERIOD) {
+                // The event is a late decryption, so store the time it took.
+                // If the time to decrypt is longer than
+                // MAXIMUM_LATE_DECRYPTION_PERIOD, we consider the event as
+                // undecryptable, and leave timeToDecryptMillis undefined
+                failure.timeToDecryptMillis = timeToDecryptMillis;
+            }
+            this.failuresToReport.add(failure);
+        }
+    }
+
+    private async handleKeysChanged(client: MatrixClient): Promise<void> {
+        if (this.checkingVerificationStatus) {
+            this.retryVerificationStatus = true;
+        } else {
```

Comment:
```suggestion
        if (this.checkingVerificationStatus) {
            // Flag that we'll need to do another check once the current check completes.
            this.retryVerificationStatus = true;
            return;
        }
```

... and remove some indentation below.